hybrisprovider.cpp

```diff
@@ -593,74 +256,41 @@ HybrisProvider::HybrisProvider(QObject *parent)
         }
     }
 
-    const hw_module_t *hwModule;
-
-    int error = hw_get_module(GPS_HARDWARE_MODULE_ID, &hwModule);
-    if (error) {
-        qWarning("Android GPS interface not found, error %d\n", error);
-        return;
-    }
-
-    qWarning("Android GPS hardware module \"%s\" \"%s\" %u.%u\n", hwModule->id, hwModule->name,
-             hwModule->module_api_version, hwModule->hal_api_version);
-
-    error = hwModule->methods->open(hwModule, GPS_HARDWARE_MODULE_ID,
-                                    reinterpret_cast<hw_device_t **>(&m_gpsDevice));
-    if (error) {
-        qWarning("Failed to open GPS device, error %d\n", error);
-        return;
+    QDir pluginsDir(qApp->applicationDirPath());
+    bool binderBackendFound = false;
+    foreach (QString fileName, pluginsDir.entryList(QDir::Files)) {
+        if (fileName.contains("libgeoclue-hybris-binder.so")) {
+            binderBackendFound = true;
+        }
     }
 
-    m_gps = m_gpsDevice->get_gps_interface(m_gpsDevice);
-    if (!m_gps) {
-        m_status = StatusError;
-        return;
+    qWarning("Loading backend plugin.");
+    QPluginLoader pluginLoader(pluginsDir.absoluteFilePath(pluginsDir.absolutePath()+"/"+(binderBackendFound ? "libgeoclue-hybris-binder.so" : "libgeoclue-hybris-hal.so")));
+    QObject *plugin = pluginLoader.instance();
+    if (plugin) {
+        m_backend = qobject_cast<HybrisBackend *>(plugin);
+    } else {
+        qWarning(pluginLoader.errorString().toStdString().c_str());
```

Comment:
`qPrintable`, maybe? Or maybe not, it's fine either way.

Reply:
I'll have a look if this can be made simpler.